torch/_inductor/fx_passes/group_batch_fusion.py

```diff
@@ -0,0 +1,183 @@
+import collections
+import logging
+import operator
+import time
+
+import torch
+from torch._dynamo.utils import counters
+
+from .. import config
+from ..pattern_matcher import CallFunctionVarArgs
+
+try:
+    # importing this will register fbgemm lowerings for inductor
+    import deeplearning.fbgemm.fbgemm_gpu.fb.inductor_lowerings  # noqa: F401
+
+    has_fbgemm = True
+except Exception:
+    has_fbgemm = False
+    pass
+
+aten = torch.ops.aten
+
+log = logging.getLogger(__name__)
+
+maximum_group_size = 50
+
+
+def _has_path(src_node, dest_node, cache):
+    """
+    If the graph has a path from `src_node` to `dest_node`, return True
+    """
+    cached = cache[id(src_node)][id(dest_node)]
+    if cached is not None:
+        return cached
+    src_input_nodes = src_node.all_input_nodes
+    if dest_node in src_input_nodes:
+        cache[id(src_node)][id(dest_node)] = True
+        return True
+
+    for x in src_input_nodes:
+        if _has_path(x, dest_node, cache):
+            cache[id(src_node)][id(dest_node)] = True
+            return True
+
+    cache[id(src_node)][id(dest_node)] = False
+    return False
+
+
+def _get_independent_node_subsets(node_list):
+    """
+    Return an iterator of node subset, each subset only contains nodes
+    those are independent with each other.
+    """
+    cache = collections.defaultdict(lambda: collections.defaultdict(lambda: None))
+
+    while len(node_list) > 0:
+        next_round = []
+
+        independent_set = []
+
+        def is_dependent(node_set, other):
+            for node in node_set:
+                if _has_path(node, other, cache) or _has_path(other, node, cache):
+                    return True
+
+            return False
+
+        for node in node_list:
+            if len(independent_set) < maximum_group_size and not is_dependent(
+                independent_set, node
+            ):
+                independent_set.append(node)
+            else:
+                next_round.append(node)
+
+        yield independent_set
+
+        cache.clear()
+
+        node_list = next_round
+
+
+class GroupBatchFusionBase:
+    def match(self, node):
+        raise NotImplementedError("match called on base")
+
+    def fuse(self, graph, subset):
+        raise NotImplementedError("fuse called on base")
+
+
+class GroupFusion(GroupBatchFusionBase):
+    """
+    Fuse ops in a group way, e.g, fuse mm/addmm of arbitrary input shapes with fbgemm.gmm.
+    """
+
+    pass
+
+
+class BatchFusion(GroupBatchFusionBase):
+    """
+    Fuse ops in a batch way, e.g, fuse mm/addmm of same input shapes with bmm.
+    """
+
+    pass
+
+
+class GroupLinearFusion(GroupFusion):
+    def match(self, node):
+        if (
+            CallFunctionVarArgs(aten.addmm.default).match(node)
+            and len(node.args[1].meta["tensor_meta"].shape) == 2
+            and node.kwargs.get("beta", 1.0) == 1.0
+            and node.kwargs.get("alpha", 1.0) == 1.0
+        ):
```

Comment:
```aten.mm``` can be fused here as well, I'll add it after fbgemm team relaxes the limitation that group transpose should be consistent.